### DISCUSSION TITLE
Tina Admin - Content discovery with search

### DISCUSSION THREAD

**Discussion by mitchmac** [tinacms/tinacms#2299] [Ideas]
It can become tedious finding the file you would like to edit once a collection has more than a handful.

Searching by title seems like a first step to make that easier. Since what is intended to be the title in a content model does not necessarily need to be mapped to a 'title' frontmatter/field, I believe this would require some configuration or filtering logic.

~~Can this be done without a proper database?~~

^Update: Tina now indexes content in a database, so the story for providing search is now easier

**Comment by jeffsee55** [comment]
Without a database this would require that we request the content for each file, parse it out to see if it has a `title` attribute and use it so I don't think this would be viable. So yes, it saves us from having to know which content model this document came from, but doesn't mitigate the real problem

**Reply by Perni1984** [reply]
I believe supporting a search api out of the box (e.g. Algolia that you are already using for your docs) would also be a good alternative - which maybe can be implemented faster and with more features (until you have a self developed solution on par).

**Comment by Perni1984** [comment]
@mitchmac @jeffsee55 is there any movement on this topic? We are having customer inquiries on the search feature.

**Reply by jamespohalloran** [reply]
We've been chatting about it internally lately so hope to have a clearer picture of how this will work soon. We're in the midst of publishing our [public roadmap](https://github.com/tinacms/tina.io/pull/1435/files), and it's in the "Next 3 months" timeframe. 
Since the original ticket was created, we're in a much better place to handle search now that we index each project's content (and don't need to rely on GitHub's API). It still remains to be seen if we'll support search out of the box, provide a hook for implementing search with a service like Algolia, or both. 

^ A few questions for you @Perni1984 though, would you have a preference out of those 2 options (Algolia support vs Tina Cloud built-in search)? Would search on just the filename work for you? Or would your users need search on the fields, and full-text search?

**Reply by Perni1984** [reply]
@jamespohalloran for the first iteration being able to search on fields of the collection would be fine. If this is limited to indexed fields (e.g. filename as I assume) it would be fine aswell.

**Reply by appinteractive** [reply]
@Perni1984 whats the status? In the Readmap its marked as "Released" but I cant find anything on that matter.

**Reply by Perni1984** [reply]
@appinteractive it's kind of there already. 

If you choose your field as "Filter By" the search field pops up:
![image](https://user-images.githubusercontent.com/9039956/232457800-db269ab6-2326-48b5-93e1-2fba06202ea4.png)

**Reply by appinteractive** [reply]
Negative, I do not have a single option up there with the latest version of tina.

<img width="2056" alt="grafik" src="https://user-images.githubusercontent.com/432828/232474957-2a7d7624-e4cb-4d55-b708-e64a506c8b4b.png">

**Reply by jamespohalloran** [reply]
@appinteractive Do you have any root level sortable fields for this collection? (One of the following non-list types: ['string', 'number', 'datetime', 'boolean'] )

If so, would you mind sharing your collection schema?

**Reply by appinteractive** [reply]
Sure @jamespohalloran 

I currently only have `simple` `page` content. I tried to move the `title`, `description` etc. out of templates into fields as they are the same for `simple` and `block` but with no luck, the compiler screams at me. 

Also would I like to make custom properties searchable line the url that is derived from breadcrumbs. 

```
import { defineConfig } from 'tinacms'
import { heroBlock, featureBlock, contentBlock } from './blocks/blocks'
import { footer } from '../tina/collections/nav-footer'
import { main } from '../tina/collections/nav-main'
import { VideoPlayerTemplate } from '../tina/embeds/video-player'
import { ContentGalleryTemplate } from './embeds/content-gallery'
import { ImageGalleryTemplate } from './embeds/image-gallery'

// Your hosting provider likely exposes this as an environment variable
const branch = process.env.HEAD || process.env.VERCEL_GIT_COMMIT_REF || 'main'

const menuItem = {
  title: {
    type: 'string',
    name: 'title',
    label: 'Titel',
    description: 'Titel der Seite (erscheint im Browser-Tab)',
    isTitle: true,
    required: true,
  },
  description: {
    type: 'string',
    name: 'description',
    label: 'Beschreibung',
    description: 'Kurze Beschreibung für Suchmaschinen',
    ui: {
      component: 'textarea',
    },
    // required: true,
  },
  teaser: {
    type: 'image',
    name: 'teaser',
    label: 'Beitragsbild',
    description: 'Bild für Teaser und Social Media',
  },
  alias: {
    type: 'string',
    name: 'alias',
    label: 'Alte URL(s)',
    description: 'Umleitung von alten URLs, beginnend mit Slash',
    list: true,
  },
  accessible: {
    type: 'string',
    name: 'accessible',
    label: 'Leichte Sprache',
    description: 'URL der Seite in leichter Sprach, beginnend mit Slash',
  },
} as any

export default defineConfig({
  branch,
  clientId: '',
  token: '',
  build: {
    outputFolder: 'admin',
    publicFolder: 'public',
  },
  media: {
    tina: {
      mediaRoot: 'media',
      publicFolder: 'public',
    },
  },
  schema: {
    collections: [
      {
        name: 'page',
        label: 'Seiten',
        path: 'content/pages',
        format: 'mdx',
        templates: [
          {
            name: 'simple',
            label: 'Einfach',
            fields: [
              menuItem.title,
              menuItem.description,
              menuItem.teaser,
              menuItem.alias,
              menuItem.accessible,
              {
                type: 'rich-text',
                name: 'body',
                label: 'Inhalt',
                isBody: true,
                templates: [
                  VideoPlayerTemplate,
                  ContentGalleryTemplate,
                  ImageGalleryTemplate,
                ],
              },
            ],
          },
          {
            name: 'blocks',
            label: 'Blöcke',
            fields: [
              menuItem.title,
              menuItem.description,
              menuItem.teaser,
              menuItem.alias,
              menuItem.accessible,
              {
                type: 'object',
                list: true,
                name: 'blocks',
                label: 'Sections',
                ui: {
                  visualSelector: true,
                },
                templates: [heroBlock, featureBlock, contentBlock],
              },
            ],
          },
        ],
        ui: {
          router: ({ document }) => {
            return `/${document._sys.breadcrumbs.join('/')}`
          },
        },
      },
      {
        name: 'navMain',
        label: 'Hauptnavigation',
        path: 'config/navigation',
        format: 'json',
        templates: [main],
        match: {
          include: 'main',
        },
        ui: {
          global: true,
          allowedActions: {
            create: false,
            delete: false,
          },
        },
      },
      {
        name: 'navFooter',
        label: 'Footer',
        path: 'config/navigation',
        format: 'json',
        templates: [footer],
        match: {
          include: 'footer',
        },
        ui: {
          global: true,
          allowedActions: {
            create: false,
            delete: false,
          },
        },
      },
    ],
  },
})

```

**Reply by jeffsee55** [reply]
@appinteractive I can confirm that filtering isn't currently supported for collections with `templates`. Curious what the errors are that you're getting when switching to `fields`.

**Reply by appinteractive** [reply]
I mean that I can't use fields and templates at the same time.

<img width="982" alt="grafik" src="https://user-images.githubusercontent.com/432828/232847043-a75d83ca-0b99-458b-bc11-be65ec9217c3.png">

**Reply by appinteractive** [reply]
so is there any workaround here? Would it be an idea to specify which fields can be searched and/or filtered?

**Reply by jeffsee55** [reply]
There isn't, though the only thing blocking this is the UI for it. I just made a ticket for this https://github.com/tinacms/tinacms/issues/3829

**Comment by danielmgzzg** [comment]
How about [Fuse.js?](https://fusejs.io/)
